llvm/unittests/Telemetry/TelemetryTest.cpp

```diff
@@ -0,0 +1,707 @@
+
+private:
+  // Returns a copy of the given entry, but with some fields sanitized.
+  CustomTelemetryEvent sanitizeFields(const CustomTelemetryEvent *Entry) {
+    CustomTelemetryEvent Sanitized(*Entry);
+    // Pretend that messages stored at EVEN positions are "sensitive",
+    // hence need to be sanitized away.
+    int S = Sanitized.Msgs.size() - 1;
+    for (int I = S % 2 == 0 ? S : S - 1; I >= 0; I -= 2)
+      Sanitized.Msgs[I] = "";
+
+    return Sanitized;
+  }
+
+  llvm::Error SendToBlackbox(json::Object O) {
+    // Here is where the vendor-defined Destination class can
+    // send the data to some internal storage.
+    // For testing purposes, we just queue up the entries to
+    // the vector for validation.
+    CurrentContext->EmittedJsons.push_back(std::move(O));
+    return Error::success();
+  }
+  bool ShouldSanitize;
+};
+
+// Custom vendor-defined Telemeter that has additional data-collection point.
+class TestTelemeter : public Telemeter {
+public:
+  TestTelemeter(std::string SessionId) : Uuid(SessionId), Counter(0) {}
+
+  static std::unique_ptr<TestTelemeter> createInstance(Config *config) {
+    if (!config->EnableTelemetry)
+      return nullptr;
+    CurrentContext->ExpectedUuid = nextUuid();
+    std::unique_ptr<TestTelemeter> Telemeter =
+        std::make_unique<TestTelemeter>(CurrentContext->ExpectedUuid);
+    // Set up Destination based on the given config.
+    for (const std::string &Dest : config->AdditionalDestinations) {
+      // The destination(s) are ALSO defined by vendor, so it should understand
+      // what the name of each destination signifies.
+      if (Dest == JSON_DEST.str()) {
+        Telemeter->addDestination(new vendor_code::JsonStreamDestination(
+            CurrentContext->SanitizeData));
+      } else if (Dest == STRING_DEST.str()) {
+        Telemeter->addDestination(new vendor_code::StringDestination(
+            CurrentContext->SanitizeData, CurrentContext->Buffer));
+      } else {
+        llvm_unreachable(
+            llvm::Twine("unknown destination: ", Dest).str().c_str());
+      }
+    }
+    return Telemeter;
+  }
+
+  void atStartup(llvm::StringRef ToolPath, TelemetryInfo *Entry) override {
+    ToolName = ToolPath.str();
+
+    // The vendor can add additional stuff to the entry before logging.
+    if (auto *S = dyn_cast<StartupEvent>(Entry)) {
+      S->MagicStartupMsg = llvm::Twine("Startup_", ToolPath).str();
+    }
+    emitToDestinations(Entry);
+  }
+
+  void atExit(llvm::StringRef ToolPath, TelemetryInfo *Entry) override {
+    // Ensure we're shutting down the same tool we started with.
+    if (ToolPath != ToolName) {
+      std::string Str;
+      raw_string_ostream OS(Str);
+      OS << "Expected tool with name" << ToolName << ", but got " << ToolPath;
+      llvm_unreachable(Str.c_str());
+    }
+
+    // The vendor can add additional stuff to the entry before logging.
+    if (auto *E = dyn_cast<ExitEvent>(Entry)) {
```

Comment:
Should the function maybe be taking a `ExitEvent` to begin with? (So that one can't even write something like `atExit(new StartupEvent)`)

Reply:
Then the base `Telemeter` does not have any method decl at all?

Reply:
I think this goes back to the question of how much can different users actually share.

We have two camps here:
 - One: nothing much (which is why we didn't actually need common API in LLVM to begin with)
 - Two: some common behaviours (eg., startup, exit)

It seems the decision was to go with two. (I'm kind of in camp ONE since the beginning, but if we've decided to go with TWO, then there should at least some common methods in the base class. Otherwise it looks pointless).

WDYT?

Reply:
I think you've summarized the problem very well, and I think I am in the first camp. I don't really have much to add here (:shrug:), so don't let this block you.